core/src/main/java/com/google/errorprone/bugpatterns/EmptyCatch.java

```diff
@@ -0,0 +1,372 @@
+ * "Simple Testing Can Prevent Most Critical Failures: 
+ *  An Analysis of Production Failures in Distributed Data-intensive Systems"
+ *  Yuan et al. Proceedings of the 11th USENIX Symposium on Operating Systems Design 
+ *  and Implementation (OSDI), 2014
+ *
+ * @author yuan@eecg.utoronto.ca (Ding Yuan)
+ */
+@BugPattern(name = "EmptyCatch",
+    summary = "Exception handling block is empty",
+    explanation = "It is rarely correct to do nothing in response to a caught exception. In production systems, "
+            + "Murphy's law often applies: Anything that can go wrong, will go wrong. "
+            + "Therefore it is particularly important to properly handle the exceptions "
+            + "regardless how rarely they are likely to occur.\n\n"
+            + "Read Effective Java, Item 65, \"Don't ignore exceptions\"; "
+            + "[Google Java Style guide section 6.2] "
+            + "(https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions); and "
+            + "\"[Simple Testing Can Prevent Most Critical Failures] "
+            + "(http://www.eecg.toronto.edu/~yuan/papers/failure_analysis_osdi14.pdf)\" "
+            + "for more detailed discussions on the harm of swallowing the exceptions. ",
+    category = JDK, maturity = EXPERIMENTAL, severity = WARNING)
+public class EmptyCatch extends BugChecker implements TryTreeMatcher {
+  /* If one of the following exceptions is ignored, we suppress the warning. */
+  ImmutableSet<String> harmlessExceptions = ImmutableSet.of("java.io.FileNotFoundException", 
+    "java.lang.ReflectiveOperationException", "java.lang.ClassNotFoundException",
+    "java.lang.reflect.InvocationTargetException", "java.lang.NoSuchFieldException",
+    "java.lang.NoSuchMethodException", "java.lang.InstantiationException");
+  
+  /* If the method is one of the following, we allow it to ignore exceptions. */
+  String[] whitelistedMethods = {
+    "cleanup", "dispose", "shutdown", "close", "stop"
+  };
+
+  @Override
+  public Description matchTry (TryTree tree, VisitorState state) {
+    if (badEmptyCatchBlock(tree, state)) {
+      // If it has finally block, assume it's OK
+      BlockTree bt = tree.getFinallyBlock();
+      if (bt == null || bt.getStatements().size() == 0) {
+        CatchTree lastCatch = tree.getCatches().get(tree.getCatches().size() - 1);
+        return describeMatch(lastCatch);
+
+      }
+    }
+    return Description.NO_MATCH;
+  }
+
+  private boolean badEmptyCatchBlock(TryTree tryTree, VisitorState state) {    
+    List<? extends CatchTree> catchList = tryTree.getCatches();
+    if (catchList == null || catchList.size() == 0) {
+      // TODO: this try block does not have a catch, we should further check the 
+      // finally block!
+      return false;
+    }
+    CatchTree catchTree = catchList.get(catchList.size() - 1);
+
+    /* If there are multiple catch blocks with a try tree, we only check the last catch block. */
```

Comment:
Good point. I think we should match all of them instead of the last one (we had this heuristic at the beginning simply to reduce the number of warnings aspirator produces).